File: internal/storage/storage.go
package storage

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"sync"
)

type LogEntry struct {
	Index   int64  `json:"index"`
	Term    int64  `json:"term"`
	Command string `json:"command"` // e.g. "PUT:key:value" or "DELETE:key"
}

type Storage struct {
	mu      sync.RWMutex
	kv      map[string]string
	walPath string
	walFile *os.File
	log     []LogEntry // in-memory log of all entries (committed + uncommitted)
	commit  int64      // last committed index
}

func NewStorage(walPath string) (*Storage, error) {
	s := &Storage{
		kv:      make(map[string]string),
		walPath: walPath,
		log:     make([]LogEntry, 0),
		commit:  0,
	}

	// Always ensure there is a dummy entry at index 0 with term 0
	// This simplifies Raft 1-based indexing logic and prevLogIndex checks.
	s.log = append(s.log, LogEntry{Index: 0, Term: 0, Command: ""})

	if err := s.replayWAL(); err != nil {
		return nil, fmt.Errorf("failed to replay WAL: %w", err)
	}

	// Open WAL for append-only writing
	f, err := os.OpenFile(walPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		return nil, fmt.Errorf("failed to open WAL file: %w", err)
	}
	s.walFile = f

	return s, nil
}

// replayWAL reads the WAL file from disk and reconstructs the log and state machine map.
func (s *Storage) replayWAL() error {
	f, err := os.OpenFile(s.walPath, os.O_RDONLY, 0666)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if len(strings.TrimSpace(line)) == 0 {
			continue
		}

		var entry LogEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			slog.Warn("Skipping corrupt WAL entry", "line", line, "error", err)
			continue
		}

		// Rebuild in-memory log
		// Pad log if indices are non-contiguous or if index is larger than current length
		for int64(len(s.log)) <= entry.Index {
			s.log = append(s.log, LogEntry{})
		}
		s.log[entry.Index] = entry

		// Apply to state machine map
		s.applyToMap(entry.Command)

		if entry.Index > s.commit {
			s.commit = entry.Index
		}
	}

	return scanner.Err()
}

// applyToMap parses the command and applies it to the in-memory map.
func (s *Storage) applyToMap(cmd string) {
	if cmd == "" {
		return
	}
	parts := strings.SplitN(cmd, ":", 3)
	if len(parts) < 2 {
		return
	}
	op := parts[0]
	key := parts[1]

	switch op {
	case "PUT":
		if len(parts) == 3 {
			s.kv[key] = parts[2]
		}
	case "DELETE":
		delete(s.kv, key)
	}
}

// Close closes the WAL file descriptor.
func (s *Storage) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.walFile != nil {
		return s.walFile.Close()
	}
	return nil
}

// AppendAndApply appends a committed log entry to the WAL file, fsyncs it, and applies it to the state machine.
func (s *Storage) AppendAndApply(entry LogEntry) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Ensure the log contains this entry in memory
	for int64(len(s.log)) <= entry.Index {
		s.log = append(s.log, LogEntry{})
	}
	s.log[entry.Index] = entry

	// Write to disk WAL
	data, err := json.Marshal(entry)
	if err != nil {
		return fmt.Errorf("failed to marshal log entry: %w", err)
	}

	if _, err := s.walFile.Write(append(data, '\n')); err != nil {
		return fmt.Errorf("failed to write to WAL: %w", err)
	}

	// fsync after every append
	if err := s.walFile.Sync(); err != nil {
		return fmt.Errorf("failed to fsync WAL: %w", err)
	}

	// Apply to in-memory KV map
	s.applyToMap(entry.Command)

	if entry.Index > s.commit {
		s.commit = entry.Index
	}

	return nil
}

// Get returns the value of a key from the state machine.
func (s *Storage) Get(key string) (string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	val, ok := s.kv[key]
	return val, ok
}

// GetLog returns a copy of the in-memory log starting from the given index (inclusive).
func (s *Storage) GetLog(fromIndex int64) []LogEntry {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if fromIndex < 0 || fromIndex >= int64(len(s.log)) {
		return nil
	}
	copied := make([]LogEntry, len(s.log)-int(fromIndex))
	copy(copied, s.log[fromIndex:])
	return copied
}

// AppendInMemory appends uncommitted entries to the in-memory log (for followers or leader log replication).
// It truncates any conflicting entries at or after the entry's index.
func (s *Storage) AppendInMemory(entries []LogEntry) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for _, entry := range entries {
		if entry.Index < 1 {
			continue
		}
		// If there is a conflict (same index, different term), truncate the log from this index onwards
		if entry.Index < int64(len(s.log)) {
			if s.log[entry.Index].Term != entry.Term {
				s.log = s.log[:entry.Index]
			} else {
				// Already exists and matches term, skip appending
				continue
			}
		}

		// Extend log if needed
		for int64(len(s.log)) < entry.Index {
			s.log = append(s.log, LogEntry{})
		}
		if int64(len(s.log)) == entry.Index {
			s.log = append(s.log, entry)
		} else {
			s.log[entry.Index] = entry
		}
	}
}

// LastLogInfo returns the index and term of the last entry in the in-memory log.
func (s *Storage) LastLogInfo() (int64, int64) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if len(s.log) == 0 {
		return 0, 0
	}
	last := s.log[len(s.log)-1]
	return last.Index, last.Term
}

// GetEntry returns the entry at the given index from the in-memory log.
func (s *Storage) GetEntry(index int64) (LogEntry, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if index < 0 || index >= int64(len(s.log)) {
		return LogEntry{}, false
	}
	return s.log[index], true
}

// CommitIndex returns the last committed index.
func (s *Storage) CommitIndex() int64 {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.commit
}

// LogLength returns the total number of entries in the in-memory log.
func (s *Storage) LogLength() int64 {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return int64(len(s.log))
}

File: internal/storage/storage.py
import os
import json
import logging
import threading
from typing import List, Tuple, Dict, Optional

class LogEntry:
    def __init__(self, index: int = 0, term: int = 0, command: str = ""):
        self.index = index
        self.term = term
        self.command = command

    def to_dict(self) -> dict:
        return {
            "index": self.index,
            "term": self.term,
            "command": self.command,
        }

    @classmethod
    def from_dict(cls, d: dict) -> 'LogEntry':
        return cls(
            index=d.get("index", 0),
            term=d.get("term", 0),
            command=d.get("command", ""),
        )

    def __repr__(self) -> str:
        return f"LogEntry(index={self.index}, term={self.term}, command='{self.command}')"


class Storage:
    def __init__(self, wal_path: str):
        self.mu = threading.RLock()
        self.kv: Dict[str, str] = {}
        self.wal_path = wal_path
        self.log: List[LogEntry] = []
        self.commit: int = 0

        # Always ensure there is a dummy entry at index 0 with term 0
        self.log.append(LogEntry(index=0, term=0, command=""))

        # Replay WAL to restore state
        self._replay_wal()

        # Open WAL for append-only writing
        self.wal_file = open(wal_path, "a", encoding="utf-8")

    def _replay_wal(self):
        if not os.path.exists(self.wal_path):
            return

        with open(self.wal_path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    data = json.loads(line)
                    entry = LogEntry.from_dict(data)
                except Exception as e:
                    logging.warning("Skipping corrupt WAL entry: %s, error: %s", line, e)
                    continue

                # Rebuild in-memory log
                while len(self.log) <= entry.index:
                    self.log.append(LogEntry())
                self.log[entry.index] = entry

                # Apply to state machine map
                self._apply_to_map(entry.command)

                if entry.index > self.commit:
                    self.commit = entry.index

    def _apply_to_map(self, cmd: str):
        if not cmd:
            return
        parts = cmd.split(":", 2)
        if len(parts) < 2:
            return
        op = parts[0]
        key = parts[1]

        if op == "PUT":
            if len(parts) == 3:
                self.kv[key] = parts[2]
        elif op == "DELETE":
            self.kv.pop(key, None)

    def close(self):
        with self.mu:
            if hasattr(self, 'wal_file') and self.wal_file:
                self.wal_file.close()

    def append_and_apply(self, entry: LogEntry) -> None:
        with self.mu:
            # Ensure the log contains this entry in memory
            while len(self.log) <= entry.index:
                self.log.append(LogEntry())
            self.log[entry.index] = entry

            # Write to disk WAL
            line = json.dumps(entry.to_dict(), separators=(',', ':')) + "\n"
            self.wal_file.write(line)
            self.wal_file.flush()
            try:
                os.fsync(self.wal_file.fileno())
            except Exception as e:
                # Some environments might not support fsync on certain file descriptors
                pass

            # Apply to in-memory KV map
            self._apply_to_map(entry.command)

            if entry.index > self.commit:
                self.commit = entry.index

    def get(self, key: str) -> Tuple[str, bool]:
        with self.mu:
            if key in self.kv:
                return self.kv[key], True
            return "", False

    def get_log(self, from_index: int) -> List[LogEntry]:
        with self.mu:
            if from_index < 0 or from_index >= len(self.log):
                return []
            # Return a copy to avoid concurrency/modification issues
            return [LogEntry(e.index, e.term, e.command) for e in self.log[from_index:]]

    def append_in_memory(self, entries: List[LogEntry]) -> None:
        with self.mu:
            for entry in entries:
                if entry.index < 1:
                    continue
                # If there is a conflict (same index, different term), truncate the log from this index onwards
                if entry.index < len(self.log):
                    if self.log[entry.index].term != entry.term:
                        self.log = self.log[:entry.index]
                    else:
                        # Already exists and matches term, skip appending
                        continue

                # Extend log if needed
                while len(self.log) < entry.index:
                    self.log.append(LogEntry())
                if len(self.log) == entry.index:
                    self.log.append(entry)
                else:
                    self.log[entry.index] = entry

    def last_log_info(self) -> Tuple[int, int]:
        with self.mu:
            if not self.log:
                return 0, 0
            last = self.log[-1]
            return last.index, last.term

    def get_entry(self, index: int) -> Tuple[Optional[LogEntry], bool]:
        with self.mu:
            if index < 0 or index >= len(self.log):
                return None, False
            return self.log[index], True

    def commit_index(self) -> int:
        with self.mu:
            return self.commit

    def log_length(self) -> int:
        with self.mu:
            return len(self.log)
